File: agent-driven-architecture-main/examples/kotlin/block/console/src/main/kotlin/adr/blocks/console/Fold.kt
// ── blocks/console/fold — the block's ARM ──────────────────────────────────
// A presentation arm obeys the SAME three rules as a domain arm (§7): read state,
// gate the effect on success, reject per-item. It has no effects at all — a
// presentation decision changes what the system believes, not what the world sees —
// so rule 2 is satisfied vacuously and rule 1 and 3 are doing real work.

package adr.blocks.console

import adr.contract.ConsoleResult
import adr.spine.pure.ArmOut
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp

internal class ConsoleArm {

    fun arm(
        slice: ConsoleSlice,
        result: ConsoleResult,
        now: Timestamp,
        sig: Signature,
    ): ArmOut<ConsoleSlice> = when (result) {
        is ConsoleResult.FocusTicket ->
            if (slice.focused == result.ticket) {
                ArmOut(
                    slice = slice,
                    notices = listOf(
                        Notice.Rejected(
                            now,
                            result.tool,
                            "ticket ${result.ticket.value} is already focused",
                        ),
                    ),
                )
            } else {
                ArmOut(slice = slice.withFocus(result.ticket))
            }

        is ConsoleResult.SetPanel ->
            if (!slice.panels.containsKey(result.panel)) {
                ArmOut(
                    slice = slice,
                    notices = listOf(
                        Notice.Rejected(now, result.tool, "no panel named ${result.panel.value}"),
                    ),
                )
            } else {
                ArmOut(slice = slice.withPanel(result.panel, result.visible))
            }
    }
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/console/fold.ts
// ── blocks/console/fold — a presentation ARM obeys the same three rules ────
// It reads current state, it folds a per-item Notice when it refuses, and it
// never touches the session-global RunStatus. Nothing about presentation earns
// a discount.

import type { Signature } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";
import { rejected } from "@adr/spine/pure/notice";
import type { ArmOut } from "@adr/spine/pure/verb";
import { armOut } from "@adr/spine/pure/verb";
import type { ConsoleResult } from "./contract";
import type { ConsoleSlice } from "./slice";
import { knowsPanel, withFocus, withPanel } from "./slice";

export function consoleArm(
  slice: ConsoleSlice,
  r: ConsoleResult,
  now: Timestamp,
  _sig: Signature,
): ArmOut<ConsoleSlice> {
  switch (r.tool) {
    case "focusTicket":
      return armOut(withFocus(slice, r.ticket), [], []);
    case "setPanel": {
      if (!knowsPanel(slice, r.panel)) {
        return armOut(slice, [], [rejected(now, r.tool, `unknown panel ${r.panel}`)]);
      }
      return armOut(withPanel(slice, r.panel, r.visible), [], []);
    }
    default: {
      const _never: never = r;
      return _never;
    }
  }
}
